File: common/src/main/java/com/jon/common/cot/UtcTimestamp.java
package com.jon.common.cot;

import androidx.annotation.NonNull;

import org.threeten.bp.Instant;

import java.util.concurrent.TimeUnit;

/* For "easier" translating between milliseconds and strings without littering annoying little
 * utility functions all over the place. */
public class UtcTimestamp {
    private String isoTimestamp;
    private long milliseconds;

    public long toLong() {
        return milliseconds;
    }

    @NonNull
    @Override
    public String toString() {
        return isoTimestamp;
    }

    public static UtcTimestamp now() {
        return new UtcTimestamp(System.currentTimeMillis());
    }

    public UtcTimestamp(String str) {
        milliseconds = stringToLong(str);
        isoTimestamp = str;
    }

    public UtcTimestamp(long ms) {
        milliseconds = ms;
        isoTimestamp = longToString(ms);
    }

    public UtcTimestamp add(long differenceMilliseconds) {
        return new UtcTimestamp(milliseconds + differenceMilliseconds);
    }

    public UtcTimestamp add(long dt, TimeUnit timeUnit) {
        return add(timeUnit.toMillis(dt));
    }

    private long stringToLong(String timestamp) {
        return Instant.parse(timestamp).toEpochMilli();
    }

    private String longToString(long ms) {
        return Instant.ofEpochMilli(ms).toString();
    }
}

File: common/src/main/java/com/jon/common/cot/UtcTimestamp.kt
package com.jon.common.cot

import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.util.concurrent.TimeUnit

/* For easier translating between milliseconds and ISO timestamp strings. */
class UtcTimestamp(private var milliseconds: Long) {

    constructor(timestampString: String) : this(stringToLong(timestampString))

    fun milliseconds(): Long = milliseconds

    fun isoTimestamp(): String = longToString(milliseconds)

    fun shortTimestamp(): String {
        /* For an ISO string of "1970-01-02Z03:04:56.111", returns "03:04:56" */
        return SHORT_FORMATTER.format(Instant.ofEpochMilli(milliseconds))
    }

    fun add(differenceMilliseconds: Long): UtcTimestamp = UtcTimestamp(milliseconds + differenceMilliseconds)

    fun add(dt: Long, timeUnit: TimeUnit): UtcTimestamp = add(timeUnit.toMillis(dt))

    companion object {
        val SHORT_FORMATTER = DateTimeFormatter
                .ofPattern("HH:mm:ss")
                .withZone(ZoneOffset.UTC)

        fun now(): UtcTimestamp = UtcTimestamp(System.currentTimeMillis())

        private fun stringToLong(timestamp: String): Long = Instant.parse(timestamp).toEpochMilli()

        private fun longToString(ms: Long): String = Instant.ofEpochMilli(ms).toString()
    }
}
